clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+    AllArgsValid = false;
+    Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type) << ArgTy;
+  }
+  void handleSyclStructType(const FieldDecl *FD, QualType ArgTy) {
+    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams &&
+        !ArgTy->isStandardLayoutType()) {
+      AllArgsValid = false;
+      Diag.Report(FD->getLocation(), diag::err_sycl_non_std_layout_type)
+          << ArgTy;
+    } else {
+      CXXRecordDecl *RD = ArgTy->getAsCXXRecordDecl();
+      if (!RD->hasTrivialCopyConstructor()) {
+        AllArgsValid = false;
+        Diag.Report(FD->getLocation(),
+                    diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 0 << ArgTy;
+      } else if (!RD->hasTrivialDestructor()) {
+        AllArgsValid = false;
+        Diag.Report(FD->getLocation(),
+                    diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 1 << ArgTy;
+      }
+    }
+  }
+};
+
+// A type to Create and own the FunctionDecl for the kernel.
+class SYCLKernelHeader : public SyclKernelFieldHandler<SYCLKernelHeader> {
+  SYCLKernelFieldChecker &ArgChecker;
+  FunctionDecl *KernelObj = nullptr;
+
+public:
+  // TODO: More params necessary for the kernel obj header.
+  SYCLKernelHeader(Sema &S, SYCLKernelFieldChecker &ArgChecker)
+      : SyclKernelFieldHandler(S), ArgChecker(ArgChecker) {}
+  ~SYCLKernelHeader() {
+    // TODO: Should 'body' do this?  Does it need to do stuff in its destructor?
+    if (KernelObj && ArgChecker.isValid())
+      SemaRef.addSyclDeviceDecl(KernelObj);
+  }
+};
+
+class SYCLKernelBody : public SyclKernelFieldHandler<SYCLKernelBody> {
+  SYCLKernelHeader &Header;
+
+public:
+  SYCLKernelBody(Sema &S, SYCLKernelHeader &H)
+      : SyclKernelFieldHandler(S), Header(H) {}
+};
+}
+
+template<typename ... Handlers>
+void VisitKernelFields(RecordDecl::field_range Fields, Handlers& ... handlers) {
+
+  // Implements the 'for-each-visitor'  pattern.
+#define KF_FOR_EACH(FUNC)                                                      \
+  (void)std::initializer_list<int>{(handlers.FUNC(Field, ArgTy), 0)...}
+
+  for (const auto &Field : Fields){
+    QualType ArgTy = Field->getType();
+
+    if (Util::isSyclAccessorType(ArgTy))
+      KF_FOR_EACH(handleSyclAccessorType);
+    else if (Util::isSyclSamplerType(ArgTy))
+      KF_FOR_EACH(handleSyclSamplerType);
+    else if (Util::isSyclSpecConstantType(ArgTy))
+      KF_FOR_EACH(handleSyclSpecConstantType);
+    else if (ArgTy->isStructureOrClassType())
+      KF_FOR_EACH(handleStructType);
+    else if (ArgTy->isReferenceType())
+      KF_FOR_EACH(handleReferenceType);
+    else if (ArgTy->isPointerType())
+      KF_FOR_EACH(handlePointerType);
+    else if (ArgTy->isArrayType())
+      KF_FOR_EACH(handleArrayType);
+    else if (ArgTy->isScalarType())
+      KF_FOR_EACH(handleScalarType);
+    else
+      llvm_unreachable("Unsupported kernel parameter type");
```

Comment:
This almost definitely shouldn't be here.  I should add a "handleOtherType" that is only implemented by the diagnostics, which simply diagnoses with an error.  Everyone else can ignore it.